contracts/Insurance.sol

```diff
@@ -15,122 +14,112 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
+        InsurancePoolToken _token = new InsurancePoolToken("Tracer Pool Token", "TPT");
+        token = address(_token);
+        collateralAsset = tracer.tracerBaseToken();
+        poolAmount = 0;
+
+        // emit InsurancePoolDeployed(_tracer, tracer.tracerBaseToken());
     }
 
     /**
      * @notice Allows a user to stake to a given tracer market insurance pool
      * @dev Mints amount of the pool token to the user
      * @param amount the amount of tokens to stake
-     * @param market the address of the tracer market to provide insurance
      */
-    function stake(uint256 amount, address market) external override {
-        StakePool storage pool = pools[market];
-        IERC20 token = IERC20(pool.collateralAsset);
-        require(supportedTracers[market], "INS: Tracer not supported");
+    function stake(uint256 amount) external override {
+        require(perpsFactory.validTracers(address(tracer)), "Pool not deployed by perpsFactory");
 
-        token.transferFrom(msg.sender, address(this), amount);
+        IERC20 collateralToken = IERC20(collateralAsset);
+        collateralToken.transferFrom(msg.sender, address(this), amount);
         // Update pool balances and user
-        InsurancePoolToken poolToken = InsurancePoolToken(pool.token);
+        InsurancePoolToken poolToken = InsurancePoolToken(token);
         uint256 tokensToMint;
         if (poolToken.totalSupply() == 0) {
             // Mint at 1:1 ratio if no users in the pool
             tokensToMint = amount;
         } else {
             // Mint at the correct ratio =
-            //          Pool tokens (the ones to be minted) / pool.amount (the collateral asset)
+            //          Pool tokens (the ones to be minted) / poolAmount (the collateral asset)
             // Note the difference between this and withdraw. Here we are calculating the amount of tokens
             // to mint, and `amount` is the amount to deposit.
-            uint256 tokensToCollatRatio = (poolToken.totalSupply() * SAFE_TOKEN_MULTIPLY) / pool.amount;
+            uint256 tokensToCollatRatio = (poolToken.totalSupply() * SAFE_TOKEN_MULTIPLY) / poolAmount;
             tokensToMint = (tokensToCollatRatio * amount) / SAFE_TOKEN_MULTIPLY;
         }
         // Margin tokens become pool tokens
         poolToken.mint(msg.sender, tokensToMint);
-        pool.amount = pool.amount + amount;
-        emit InsuranceDeposit(market, msg.sender, amount);
+        poolAmount = poolAmount + amount;
+        emit InsuranceDeposit(address(tracer), msg.sender, amount);
     }
 
     /**
      * @notice Allows a user to withdraw their assets from a given insurance pool
      * @dev burns amount of tokens from the pool token
      * @param amount the amount of pool tokens to burn
-     * @param market the tracer contract that the insurance pool is for.
      */
-    function withdraw(uint256 amount, address market) external override {
+    function withdraw(uint256 amount) external override {
```

Comment:
Just a general thought: Should `withdraw` call `updatePoolAmount`? I realise the user can just call it themself before withdrawing, but is this extra step desired?
@raymogg

Reply:
I think it would help mitigate potential errors since it would definitely need to be there before `withdraw` is called, so I think so :+1:

Reply:
@BenjaminPatch @kumar-ish good point. It should probably also be called on deposit then hey? Otherwise ratios will be out

Reply:
Yes, it should -- fixed now